system/lib/emmalloc.cpp

```diff
@@ -0,0 +1,1187 @@
+    EM_ASM({ Module.print("    emmalloc.allocateRegion fixing alignment") });
+#endif
+    size_t extra = (char*)fixedPtr - (char*)ptr;
+    void* extraPtr = sbrk(extra);
+    if (extraPtr == (void*)-1) {
+      // sbrk() failed, we failed.
+#ifdef EMMALLOC_DEBUG_LOG
+      EM_ASM({ Module.print("    emmalloc.newAllocation sbrk failure") });;
+#endif
+      return NULL;
+    }
+    // Verify the sbrk() assumption, no one else should call it.
+    // If this fails, it means we also leak the previous allocation,
+    // so we don't even try to handle it.
+    assert((char*)extraPtr == (char*)ptr + sbrkSize);
+    // After the first allocation, everything must remain aligned forever.
+    assert(!lastRegion);
+    // We now have a contiguous block of memory from ptr to
+    // ptr + sbrkSize + fixedPtr - ptr = fixedPtr + sbrkSize.
+    // fixedPtr is aligned and starts a region of the right
+    // amount of memory.
+  }
+  Region* region = (Region*)fixedPtr;
+  // Apply globally
+  if (!lastRegion) {
+    assert(!firstRegion);
+    firstRegion = region;
+    lastRegion = region;
+  } else {
+    assert(firstRegion);
+    region->prev() = lastRegion;
+    lastRegion = region;
+  }
+  // Success, we have new memory
+#ifdef EMMALLOC_DEBUG_LOG
+  EM_ASM({ Module.print("    emmalloc.newAllocation success") });;
+#endif
+  region->setTotalSize(sbrkSize);
+  region->setUsed(1);
+  return region;
+}
+
+// Allocate new memory. This may reuse part of the last region, only
+// allocating what we need.
+static Region* newAllocation(size_t size) {
+#ifdef EMMALLOC_DEBUG_LOG
+  EM_ASM({ Module.print("  emmalloc.newAllocation " + $0) }, size);
+#endif
+  assert(size > 0);
+  if (lastRegion) {
+    // If the last region is free, we can extend it rather than leave it
+    // as fragmented free spce between allocated regions. This is also
+    // more efficient and simple as well.
+    if (!lastRegion->getUsed()) {
+#ifdef EMMALLOC_DEBUG_LOG
+      EM_ASM({ Module.print("    emmalloc.newAllocation extending lastRegion at " + $0) }, lastRegion);
+#endif
+      // Remove it first, before we adjust the size (which affects which list
+      // it should be in). Also mark it as used so extending it doesn't do
+      // freelist computations; we'll undo that if we fail.
+      lastRegion->setUsed(1);
+      removeFromFreeList(lastRegion);
+      if (extendLastRegion(size)) {
+        return lastRegion;
+      } else {
+        lastRegion->setUsed(0);
+        return NULL;
+      }
+    }
+  }
+  // Otherwise, get a new region.
+  return allocateRegion(size);
+}
+
+// Internal mirror of public API.
+
+static void* emmalloc_malloc(size_t size) {
+  // malloc() spec defines malloc(0) => NULL.
+  if (size == 0) return NULL;
+  // Look in the freelist first.
+  Region* region = tryFromFreeList(size);
+  if (!region) {
+    // Allocate some new memory otherwise.
+    region = newAllocation(size);
+    if (!region) {
+      // We failed to allocate, sadly.
+      return NULL;
```

Comment:
I guess this allocator should respect ABORTING_MALLOC too?

Reply:
ABORTING_MALLOC should just work, that is handled by sbrk (it will abort when it can't grow memory).